### DISCUSSION TITLE
Dumb Question - Does predict() use the same weights on all timesteps ?

### DISCUSSION THREAD

**Discussion by jiyer2016** [skfolio/skfolio#90] [Q&A] [answered]
I am just getting started with using skfolio - and this might be a dumb question.
After training the model, I notice the the dot product of the trained model weights (model.weights_) and the asset level returns (X) at each timestep equals the outout of the predict method.

So, it seems at each timestep - the exact same model weights is used in the dot product ??

My question is how is it possible to do that ? Shouldnt the weights be recomputed at each timestep considering that the price fluctations result in the weights drifting ?

I am sure I have something wrong with my understanding of how things work. Appreciate your inputs.

**Comment by HugoDelatte** [comment, marked as answer]
You are correct. The explanation is provided in this section (under the warning): https://skfolio.org/user_guide/portfolio.html#id2.

In short, for proper analysis, the data used for both `predict` and `fit` need to be homogeneous. For example, calling `predict` on the same sample as `fit` must give you the same results; otherwise, you may be introducing bias into the evaluation.

If needed, you could extend the class by adding a new `predict` method that accounts for asset drift.

**Comment by CarloNicolini** [comment]
> I am just getting started with using skfolio - and this might be a dumb question. After training the model, I notice the the dot product of the trained model weights (model.weights_) and the asset level returns (X) at each timestep equals the outout of the predict method.
> 
> So, it seems at each timestep - the exact same model weights is used in the dot product ??
> 
> My question is how is it possible to do that ? Shouldnt the weights be recomputed at each timestep considering that the price fluctations result in the weights drifting ?
> 
> I am sure I have something wrong with my understanding of how things work. Appreciate your inputs.

Here is a quick implementation of what you are looking for.
Pay attention that having asset drift (using the following method) is equivalent to rebalancing at every single step!
The function `predict_proba` computes the asset "components" such that the rebalancing given by `weights` is  valid at each step.
The function `predict` instead gives you the portfolio value, normalized at 1 at the first step (also called `equity_curve`).

```python

    def predict_proba(self, X):
        """
        Predicts the asset positions and equity curve based on
        historical price data and predefined weights.
        Here the initial portfolio value is **always** hypothesized to be $1.
        The asset positions are expressed in $, as well as the equity curve.

        Parameters
        ----------
        X : pd.DataFrame
            Historical price data with columns representing different assets and rows as timestamps.

        Returns
        -------
        pd.DataFrame
            DataFrame containing the calculated equity curve with the same structure as input 'X'.
        """
        check_is_fitted(self)
        assert_is_prices(X)

        # Convert to NumPy array for faster computation
        prices_np: np.ndarray = X.values
        # Calculate returns and fill NaN values with 0.0
        returns_np: np.ndarray = X.pct_change().dropna().values
        weights_np: np.ndarray = np.array(self.weights_)

        # Calculate position in currency units (dollars or euro for example) for the initial day
        # using dot product. This value will be normalized at the end in order to start from 1
        # unit of currency
        position_0 = prices_np[0, :].dot(weights_np)
        initial_value = prices_np[0, :].sum()

        # Initialize positions and equity curve arrays
        positions_np = np.zeros_like(prices_np)
        positions_np[0, :] = position_0

        # Vectorized computation of positions and equity curve
        np.cumprod(1.0 + returns_np, axis=0, out=positions_np[1:])
        positions_np[1:, :] *= position_0

        positions_np /= initial_value
        return pd.DataFrame(data=positions_np, index=X.index, columns=X.columns)
 ```
 
 finally for computing the value of the portfolio with continous rebalancing just sum horizontally

```python
    def predict(self, X):
        return self.predict_proba(X).sum(axis=1)
```
